asciidoc-confluence-publisher-client/src/main/java/org/sahli/asciidoc/confluence/publisher/client/ConfluencePublisher.java

```diff
@@ -187,43 +190,41 @@ private void updatePage(String contentId, String ancestorId, ConfluencePageMetad
         }
     }
 
-    private static String contentHash(String content) {
-        return sha256Hex(content);
-    }
-
     private void addAttachments(String contentId, Map<String, String> attachments) {
         attachments.forEach((attachmentFileName, attachmentPath) -> addOrUpdateAttachment(contentId, attachmentPath, attachmentFileName));
     }
 
     private void addOrUpdateAttachment(String contentId, String attachmentPath, String attachmentFileName) {
         Path absoluteAttachmentPath = absoluteAttachmentPath(attachmentPath);
+        String newContentHash = sha256Hash(fileInputStream(absoluteAttachmentPath));
 
         try {
             ConfluenceAttachment existingAttachment = this.confluenceClient.getAttachmentByFileName(contentId, attachmentFileName);
-            InputStream existingAttachmentContent = this.confluenceClient.getAttachmentContent(existingAttachment.getRelativeDownloadLink());
+            String attachmentId = existingAttachment.getId();
+            String existingContentHash = this.confluenceClient.getPropertyByKey(contentId, attachmentId);
 
-            if (!isSameContent(existingAttachmentContent, fileInputStream(absoluteAttachmentPath))) {
-                this.confluenceClient.updateAttachmentContent(contentId, existingAttachment.getId(), fileInputStream(absoluteAttachmentPath));
+            if (notSameHash(existingContentHash, newContentHash)) {
+                this.confluenceClient.deletePropertyByKey(contentId, attachmentId);
+                this.confluenceClient.updateAttachmentContent(contentId, attachmentId, fileInputStream(absoluteAttachmentPath));
+                this.confluenceClient.setPropertyByKey(contentId, attachmentId, newContentHash);
             }
 
         } catch (NotFoundException e) {
-            this.confluenceClient.addAttachment(contentId, attachmentFileName, fileInputStream(absoluteAttachmentPath));
+            ConfluenceAttachment newAttachment = this.confluenceClient.addAttachment(contentId, attachmentFileName, fileInputStream(absoluteAttachmentPath));
```

Comment:
If you would use the attachment file name as the property name (instead of the attachment id), you could avoid returning the attachment after adding it. The attachment name has to be unique per Confluence page (i.e. uploading twice an attachment with the same name to the same page would overwrite the first one).

Reply:
Reworked using attachment file name. 
Also had to add suffix to property key ( `-hash`), since spring-web used in Confluence server by default uses everything after dot as file extension.